src/hash_generator/dlNormalizer.js

```diff
@@ -0,0 +1,398 @@
+    else if (uri.indexOf('/402/') !== -1) uri = `https://id.gs1.org${uri.substring(uri.indexOf('/402/'))}`;
+    else if (uri.indexOf('/414/') !== -1) uri = `https://id.gs1.org${uri.substring(uri.indexOf('/414/'))}`;
+    else if (uri.indexOf('/417/') !== -1) uri = `https://id.gs1.org${uri.substring(uri.indexOf('/417/'))}`;
+    else if (uri.indexOf('/8003/') !== -1) uri = `https://id.gs1.org${uri.substring(uri.indexOf('/8003/'))}`;
+    else if (uri.indexOf('/8004/') !== -1) uri = `https://id.gs1.org${uri.substring(uri.indexOf('/8004/'))}`;
+    else if (uri.indexOf('/8006/') !== -1) uri = `https://id.gs1.org${uri.substring(uri.indexOf('/8006/'))}`;
+    else if (uri.indexOf('/8010/') !== -1) uri = `https://id.gs1.org${uri.substring(uri.indexOf('/8010/'))}`;
+    else if (uri.indexOf('/8017/') !== -1) uri = `https://id.gs1.org${uri.substring(uri.indexOf('/8017/'))}`;
+    else if (uri.indexOf('/8018/') !== -1) uri = `https://id.gs1.org${uri.substring(uri.indexOf('/8018/'))}`;
+  }
+
+  if (!uri.match('^https:\/\/id.gs1.org\/01\/[0-9]{14}')) {
+    if (uri.match('^https:\/\/id.gs1.org\/01\/[0-9]{13}')) {
+      uri = uri.replace('/01/', '/01/0');
+    } else if (uri.match('^https:\/\/id.gs1.org\/01\/[0-9]{12}')) {
+      uri = uri.replace('/01/', '/01/00');
+    } else if (uri.match('^https:\/\/id.gs1.org\/01\/[0-9]{8}')) {
+      uri = uri.replace('/01/', '/01/000000');
+    }
+  }
+
+  // remove cpv
+  const x = uri.substring(uri.indexOf('/22/') + 4);
+
+  // for 01/8006 only
+  if (
+    uri.match('https:\/\/id.gs1.org\/8006\/[0-9]{18}\/22\/([\x22\x27\x2D\x2E\x30-\x39\x3B-\x3F\x41-\x5A\x5F\x61-\x7A]{0,20})$') ||
+    uri.match('https:\/\/id.gs1.org\/01\/[0-9]{14}\/22\/([\x22\x27\x2D\x2E\x30-\x39\x3B-\x3F\x41-\x5A\x5F\x61-\x7A]{0,20})$')
+  ) {
+    uri = uri.substring(0, uri.indexOf('/22/')) + x.substring(x.indexOf('/'), x.length - 1);
+  }
+
+  // for 01/8006 followed by other key qualifiers
+  if (
+    uri.match('https:\/\/id.gs1.org\/8006\/[0-9]{18}\/22\/([\x2F\x22\x27\x2D\x2E\x30-\x39\x3B-\x3F\x41-\x5A\x5F\x61-\x7A]{0,20})$') ||
+    uri.match('https:\/\/id.gs1.org\/01\/[0-9]{14}\/22\/([\x2F\x22\x27\x2D\x2E\x30-\x39\x3B-\x3F\x41-\x5A\x5F\x61-\x7A]{0,20})$')
+  ) {
+    uri = uri.substring(0, uri.indexOf('/22/')) + x.substring(x.indexOf('/'));
+  }
+
+  // take only lowest ID granularity level (i.e. if serial is present, omit lot)
+  if (
+    uri.match('https:\/\/id.gs1.org\/8006\/[0-9]{18}\/10\/([\x22\x27\x2D\x2E\x30-\x39\x3B-\x3F\x41-\x5A\x5F\x61-\x7A]{0,20})\/21\/([\x22\x27\x2D\x2E\x30-\x39\x3B-\x3F\x41-\x5A\x5F\x61-\x7A]{0,20})$') ||
+    uri.match('https:\/\/id.gs1.org\/01\/([0-9]{14})\/10\/([\x22\x27\x2D\x2E\x30-\x39\x3B-\x3F\x41-\x5A\x5F\x61-\x7A]{0,20})\/21\/([\x22\x27\x2D\x2E\x30-\x39\x3B-\x3F\x41-\x5A\x5F\x61-\x7A]{0,20})$')
+  ) {
+    const y = uri.substring(uri.indexOf('/22/') + 4);
+    uri = uri.substring(0, uri.indexOf('/10/')) + y.substring(y.indexOf('/'));
+  }
+
+  if (
```

Comment:
I have no idea what this huge block of regexs does.

Reply:
it checks if the URI is now a valid one or still not. If it is still not a valid one, it returns the original one. Otherwise, it returns the updated one.

Reply:
OK, add some comments to explain this.